### PR TITLE
uftrace: Fix fread() error handling

### PR DESCRIPTION
**Pull request by gonapps**
URGENT!
#966 induces a broken state of master branch.
Currently simple command like uftrace -S scripts/strings.py ls doesn't work.

This commit fixed the problems.

From manpage:
> fread() does not distinguish between end-of-file and error, and callers
> must use feof(3) and ferror(3) to determine which occurred.

In cmds/info.c checking ret != sizeof(fha->buf) doesn't always mean that an error occured.
If the buffer is large enough, return value of fread() can be smaller than the size of the buffer.

In utils/symbol.c when while loop ends, len will be always zero.
To know an error actually occured, check the result of ferror()

Also there were compile warnings:
```console
/home/gonapps/workspace/uftrace/cmds/info.c: In function ‘fill_cmdline’:
/home/gonapps/workspace/uftrace/cmds/info.c:230:25: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  230 |  for (i = 0, p = buf; i < ret; i++, p++) {
      |                         ^
/home/gonapps/workspace/uftrace/cmds/info.c:237:22: warning: passing argument 2 of ‘json_quote’ from incompatible pointer type [-Wincompatible-pointer-types]
  237 |  p = json_quote(buf, &ret);
      |                      ^~~~
      |                      |
      |                      size_t * {aka long unsigned int *}
In file included from /home/gonapps/workspace/uftrace/utils/symbol.h:15,
                 from /home/gonapps/workspace/uftrace/uftrace.h:11,
                 from /home/gonapps/workspace/uftrace/cmds/info.c:20:
/home/gonapps/workspace/uftrace/utils/utils.h:337:8: note: expected ‘int *’ but argument is of type ‘size_t *’ {aka ‘long unsigned int *’}
  337 | char * json_quote(char *str, int *len);
      |        ^~~~~~~~~~
/home/gonapps/workspace/uftrace/cmds/info.c:241:30: warning: comparison of integer expressions of different signedness: ‘ssize_t’ {aka ‘long int’} and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
```

Fixed: #952

Signed-off-by: Byeonggon Lee <gonny952@gmail.com>

### CODE REVIEW

**Comment by honggyukim** [comment]
In addition, why don't you write the commit message as follows?
```
The following commit is a regression that breaks every normal recording.

  5cffda2 uftrace: Fix fread() error handling

Unfortunately, this change breaks normal recording as follows:

  $ uftrace t-abc
  uftrace: /home/honggyu/work/uftrace/utils/data-file.c:493:open_info_file
   ERROR: cannot read uftrace header info!

In uftrace.data/info file, cmdline line is not found at all.

The man page of fread describes as follows:

  fread() does not distinguish between end-of-file and error, and callers
  must use feof(3) and ferror(3) to determine which occurred.

In cmds/info.c checking ret != sizeof(fha->buf) doesn't always mean that
an error occured.  If the buffer is large enough, return value of fread()
can be smaller than the size of the buffer.

In utils/symbol.c when while loop ends, len will be always zero.
ferror() can be used to check if an error actually occured in this case.

There are compile warnings as well.

  cmds/info.c: In function ‘fill_cmdline’:
  cmds/info.c:230:25: warning: comparison of integer expressions of different signedness:
                      ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
    230 |  for (i = 0, p = buf; i < ret; i++, p++) {
        |                         ^
  cmds/info.c:237:22: warning: passing argument 2 of ‘json_quote’ from
                      incompatible pointer type [-Wincompatible-pointer-types]
    237 |  p = json_quote(buf, &ret);
        |                      ^~~~
        |                      |
        |                      size_t * {aka long unsigned int *}
  In file included from /home/gonapps/workspace/uftrace/utils/symbol.h:15,
                   from /home/gonapps/workspace/uftrace/uftrace.h:11,
                   from /home/gonapps/workspace/uftrace/cmds/info.c:20:
  utils/utils.h:337:8: note: expected ‘int *’ but argument is of type ‘size_t *’
                       {aka ‘long unsigned int *’}
    337 | char * json_quote(char *str, int *len);
        |        ^~~~~~~~~~
  cmds/info.c:241:30: warning: comparison of integer expressions of different signedness:
                      ‘ssize_t’ {aka ‘long int’} and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]

The warnings can simply be suppressed by changing ret variable to int type.

Fixed: #952

Signed-off-by: Byeonggon Lee <gonny952@gmail.com>
```

**Comment by namhyung** [comment]
I like the @honggyukim 's change.

**Comment by gonapps** [comment]
@honggyukim @namhyung Fixed

**Comment by namhyung** [comment]
Please remove the unnecessary type changes.

**Comment by gonapps** [comment]
@namhyung Fixed

File: cmds/info.c [hunk 1]
```diff
@@ -209,19 +210,21 @@ static int fill_cmdline(void *arg)
 	struct fill_handler_arg *fha = arg;
 	char *buf = fha->buf;
 	FILE *fp;
+	int err;
 	size_t ret;
```
Comment (namhyung) [inline cmds/info.c]: You can make `ret` as int.

File: cmds/info.c [hunk 2]
```diff
@@ -209,19 +210,21 @@ static int fill_cmdline(void *arg)
...
 	if (fp == NULL)
 		return -1;
 
 	ret = fread(buf, 1, sizeof(fha->buf), fp);
+	err = ferror(fp);
 	fclose(fp);
 
-	if (ret != sizeof(fha->buf))
-		return ret;
+	if (err || ret > INT_MAX)
```
Comment (namhyung) [inline cmds/info.c]: I think `if (!ret && err)` is better.

Comment (gonapps) [reply cmds/info.c]: ret > INT_MAX is meant to detect fread() return value bigger than int size
Maybe I could change ret to int then check err || ret < 0

**Comment by gonapps** [comment]
Updated the commit

File: cmds/info.c (hunk 2 above)
Comment (namhyung) [reply cmds/info.c]: Hmm.. I think ret cannot be bigger than `sizeof(fha->buf)`.

Comment (gonapps) [reply cmds/info.c]: Oh you are right. I will fix as you said

**Comment by gonapps** [comment]
@namhyung 
> Hmm.. I think ret cannot be bigger than sizeof(fha->buf).
Fixed

File: cmds/info.c (hunk 2 above)
Comment (namhyung) [reply cmds/info.c]: Return value of 0 itself is not an error.  I think we should check err with `&&`.

Comment (gonapps) [reply cmds/info.c]: Thanks I'll FIx

**Comment by gonapps** [comment]
@namhyung Fixed with &&

**Pull request merged**